### ISSUE TITLE
Doesn't work in Windows

### ISSUE THREAD

**Issue by Neltherion** [ppwwyyxx/tensorpack#100]
I'm trying to run the project on Windows 10 64-bit but so far when it tries to initialize an IPC pipe, it fails with a `{ZMQError}Protocol not supported` error... 
I tried using `tcp` instead of `ipc` but it resulted in a `{ZMQError}Invalid argument` error...

Here's where the error happens : `tensorpack/dataflow/prefetch.py:157 => self.socket.bind(self.pipename)`

If I use `inproc` instead of `ipc`, this error doesn't happen but when it calls `self.start_processes()` we'll get another error with this traceback:

```
  File "PATH_TO_PACKAGE\Tensorpack\tensorpack\dataflow\prefetch.py", line 159, in __init__
    self.start_processes()
  File "PATH_TO_PACKAGE\Tensorpack\tensorpack\dataflow\prefetch.py", line 166, in start_processes
    start_proc_mask_signal(self.procs)
  File "PATH_TO_PACKAGE\Tensorpack\tensorpack\utils\concurrency.py", line 152, in start_proc_mask_signal
    p.start()
  File "C:\Anaconda3\lib\multiprocessing\process.py", line 105, in start
    self._popen = self._Popen(self)
  File "C:\Anaconda3\lib\multiprocessing\context.py", line 212, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Anaconda3\lib\multiprocessing\context.py", line 313, in _Popen
    return Popen(process_obj)
  File "C:\Anaconda3\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Anaconda3\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'AugmentImageComponents.__init__.<locals>.func'
Exception ignored in: <bound method PrefetchDataZMQ.__del__ of <tensorpack.dataflow.prefetch.PrefetchDataZMQ object at 0x000001C397C81978>>
Traceback (most recent call last):
  File "PATH_TO_PACKAGE\Tensorpack\tensorpack\dataflow\prefetch.py", line 192, in __del__
    x.terminate()
  File "C:\Anaconda3\lib\multiprocessing\process.py", line 113, in terminate
    self._popen.terminate()
AttributeError: 'NoneType' object has no attribute 'terminate'
```

has anyone been able to run this project on Windows?

**Comment by ppwwyyxx** [comment]
Windows doesn't support ipc communication: http://stackoverflow.com/questions/15386121/does-zeromq-support-ipc-as-a-transport-channel-on-windows.

To make it work you'll have to use tcp. The invalid argument error is probably because you didn't write the address correctly. It should be something like tcp://127.0.0.1:1234.
Using tcp, it probably will run as slow as the python multiprocessing module. So maybe you can just use PrefetchData instead of PrefetchDataZMQ.

**Comment by Neltherion** [comment]
@ppwwyyxx
Thanks for the quick response... Changing the protocol to `TCP` (or `inproc`) avoids the first error but the second error (which I explained in the post above) still remains... I do believe its main problem resides here:

```
  File "C:\Anaconda3\lib\multiprocessing\reduction.py", line 59, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'AugmentImageComponents.__init__.<locals>.func'
```

which eventually results in this error:

```
  File "PATH_TO_PACKAGE\Tensorpack\tensorpack\dataflow\prefetch.py", line 193, in __del__
    x.terminate()
  File "C:\Anaconda3\lib\multiprocessing\process.py", line 113, in terminate
    self._popen.terminate()
AttributeError: 'NoneType' object has no attribute 'terminate'
```

**Comment by ppwwyyxx** [comment]
inproc wouldn't work at all. Regarding the error, are you using any of the examples or have you modified the data? It shouldn't try to pickle the augmentor. Do you have an stack trace of the first error?

Btw, if you are running Image2Image.py, maybe just use `PrefetchData(ds, 100, 1)`, or just remove prefetch completely. Dataflow is not the bottleneck here.

**Comment by Neltherion** [comment]
Ok... I commented this line `ds = PrefetchDataZMQ(ds, 1)` and had to change 2 parts of the code... 

the first part was to use `config.dataflow` instead of `config.dataset` in `tensorpack-master/examples/GAN/GAN.py:18` and the second part was to change `labels` to `targets` in `sigmoid_cross_entropy_with_logits`... I'm using Tensorflow 0.12.

**Comment by ppwwyyxx** [comment]
Yes it's alright. The first one was fixed 6d67faf93a7 today. The second one is because tensorflow changes their API recently on master.

**Comment by Neltherion** [comment]
Thank you for your time... it's working right now and is pretty much faster than what [yenchenlin](https://github.com/yenchenlin/pix2pix-tensorflow) has implemented... I only need time to get around the architecture...
But somehow, this project reminds me of Keras: abstracting some layers and providing easier to work interfaces... am I wrong?
